### ISSUE TITLE
Unreference underlying socket to prevent process hangs

### ISSUE THREAD

**Issue by jsocol** [brightcove/hot-shots#57]
We recently started seeing our test suites hang across projects that were using hot-shots (we have our own wrapper which sets some defaults) and tracked it down to a new behavior in mocha 4, which prevents the test process from exiting while there are open handles.

In theory, this could also impact clean shutdowns of non-test processes as well where `StatsD` objects may not be able to be informed of a shutdown signal (for instance, if a library creates its own stats client). I have not, however, personally observed that behavior, but it seems more likely in scripts or other context that are expected to shut themselves down, rather than serve forever, e.g. a cron task.

Calling `socket.unref()` after [creating the socket](https://github.com/brightcove/hot-shots/blob/master/lib/statsd.js#L53) allows node to exit if the hot-shots socket is the last open handle. It may, however, introduce some risk of losing metrics before they're sent.

I'm not sure if this is worth doing in the library—or perhaps adding a opt-in option for—but I wanted to write it down in case other people run into it.

**Comment by jsocol** [comment]
It is pretty easy to work around, for sure. This is what we've done, since we're wrapping the `StatsD` class:

```js
class Metrics extends StatsD {
  constructor(options) {
    super(options);
    this.socket.unref();
  }
}
```

but that could also be handled by any code after instantiating a `StatsD` instance. I'm not sure how "public" `this.socket` is, but as long as it doesn't change, user-land code can fix this.

**Comment by bdeitte** [comment]
I tried to use unref() as part of https://github.com/brightcove/hot-shots/pull/89, as I ran into this issue when upgrading mocha in this project.  Oddly enough, that didn't work by itself, so I left it out for now.  I ended up having to use "--exit" as a param for mocha.  But now it should be able to test out such a fix... not sure if there is more that is needed with the tests, or something more for the codebase itself in more cases, didn't get to dig into it too much.

**Comment by bdeitte** [comment]
As noted in the issue mentioned above, there may be even worse behavior happening now in hot-shots 6.1.  Folding these issues together though because having the process not hang is a general annoyance that needs to get fixed.  @srilq does adding the --exit param work for you, or the unref mentioned above?  Just want to make sure I understand this new wrinkle as well, thanks.

**Comment by srilq** [comment]
@bdeitte 
Jest `--forceExit` works! However, this will still leave the process running in the background which isn't ideal for local development.

**Comment by danroyo** [comment]
I'm seeing some non-test scenarios where the unreferenced socket is causing a process to hang.  I'm working on a project where the parent process creates a pool of workers by forking.  Normally these child processes have a lifespan of under a minute, but we're seeing them hang because those workers send StatsD metrics and will not get torn down due to the open handle.

**Comment by benblack86** [comment]
Ran into this problem, put together this fix => https://github.com/brightcove/hot-shots/pull/167

**Comment by bdeitte** [comment]
Fixed in 7.5.0
